router/core/operation_processor.go

```diff
@@ -363,30 +448,182 @@ func (o *OperationKit) Normalize() error {
+	entry, ok := o.cache.persistedOperationCache[cacheKey]
+	o.cache.persistedOperationCacheLock.RUnlock()
+	if !ok {
+		return false, nil
+	}
+	o.parsedOperation.PersistedOperationCacheHit = true
+	o.parsedOperation.ID = entry.operationID
+	o.parsedOperation.NormalizedRepresentation = entry.normalizedRepresentation
+	o.parsedOperation.Type = entry.operationType
+	if o.parsedOperation.Request.Variables == nil || bytes.Equal(o.parsedOperation.Request.Variables, literalNull) {
+		o.parsedOperation.Request.Variables = []byte("{}")
+	}
+
+	err = jsonparser.ObjectEach(entry.exportedVariables, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) (err error) {
+		_, exists := o.parsedOperation.VariablesMap[string(key)]
+		if exists {
+			return nil
+		}
+		if dataType == jsonparser.String {
+			stringValue := make([]byte, len(value)+2)
+			stringValue[0] = '"'
+			copy(stringValue[1:], value)
+			stringValue[len(stringValue)-1] = '"'
+			o.parsedOperation.Request.Variables, err = jsonparser.Set(o.parsedOperation.Request.Variables, stringValue, unsafebytes.BytesToString(key))
+		} else {
+			o.parsedOperation.Request.Variables, err = jsonparser.Set(o.parsedOperation.Request.Variables, value, unsafebytes.BytesToString(key))
+		}
+		return
+	})
+	if err != nil {
+		return false, errors.WithStack(fmt.Errorf("failed to set default values for variables: %w", err))
+	}
+
+	o.kit.doc.Input.ResetInputString(entry.normalizedRepresentation)
+	report := &operationreport.Report{}
+	o.kit.parser.Parse(o.kit.doc, report)
+	if report.HasErrors() {
+		return false, &reportError{
+			report: report,
+		}
+	}
+
+	return true, nil
+}
+
+func (o *OperationKit) savePersistedOperationToCache(skipIncludeVariableNames []string, exportedVariables []byte) {
+	if o.cache == nil {
+		return
+	}
+	cacheKey := o.generatePersistedOperationCacheKey(skipIncludeVariableNames)
+
+	entry := normalizedOperationCacheEntry{
+		operationID:              o.parsedOperation.ID,
+		normalizedRepresentation: o.parsedOperation.NormalizedRepresentation,
+		operationType:            o.parsedOperation.Type,
+		exportedVariables:        make([]byte, len(exportedVariables)),
+	}
+	copy(entry.exportedVariables, exportedVariables)
+
+	o.cache.persistedOperationCacheLock.Lock()
+	o.cache.persistedOperationCache[cacheKey] = entry
+	o.cache.persistedOperationCacheLock.Unlock()
+
+	o.cache.persistetOperationVariableNamesLock.Lock()
+	o.cache.persistetOperationVariableNames[o.parsedOperation.GraphQLRequestExtensions.PersistedQuery.Sha256Hash] = skipIncludeVariableNames
+	o.cache.persistetOperationVariableNamesLock.Unlock()
+}
+
+var (
+	literalF    = []byte("f")
+	literalT    = []byte("t")
+	literalNull = []byte("null")
+)
+
+func (o *OperationKit) loadPersistedOperationCacheKey(persistedQuerySha256Hash string) (key uint64, ok bool) {
+	o.cache.persistetOperationVariableNamesLock.RLock()
+	variableNames, ok := o.cache.persistetOperationVariableNames[persistedQuerySha256Hash]
```

Comment:
We create an operationKit per request. From an abstraction perspective, it would be great if the kit wouldn't know anything from the cache so that the kit and the cache are not coupled. Pseudocode:

```go
kit := NewOperationKit()
normalization := kit.Normalize()
parsedOp, ok := cache.Get(sha256)
if !ok {
  kit.Parse(...)
}

```

Reply:
Yes but you cannot have a global cache here. The problem is that the overall architecture of the operation processor needs to be rewritten. You cannot really do dependency injection with the current way things are organized. Once all of this works and is well tested, I can refactor the whole pipeline. Keep in mind that the kit doesn't know anything about the cache, it's just a pointer for DI purposes. Any other suggestions to improve it without globals?

Reply:
I understand, then I'd postpone the discussion to the general refactor.